File: JavaScript/1_ProblemSolvingPatterns/2_anagramChallenge.js
/*

AnaGram Challenge is to find whether 2 Input Strings has same letters with same Frequency each.

    Example 1   =>  "aaz",   "zaa"  => True

    Exapmle 2   =>   "anagram",  "aangra"  => False


    Time Complexity for the Provided Solution is O(n)
*/

const ValidAnagram = (str1, str2) => {
  let lookUp = {};
  if (str1.length !== str2.length) return false;
  for (const val of str1) {
    lookUp[val] = ++lookUp[val] || 1;
  }

  for (let i = 0; i < str2.length; i++) {
    const element = str2[i];
    if (!lookUp[element]) {
      // if value of key in lookup dict becomes 0 this will return false
      return false;
    } else {
      lookUp[element] -= 1;
    }
  }
  return true;
};

let a = "level";
let b = "elven";
console.log(ValidAnagram(a, b));

File: Python/1_ProblemsolvingPatterns/2_anagramChallenge.py
def ChechAnagram(str1, str2):
	if len(str1) != len(str2): return False
	
	look_up = {}

	for i in str1:
		if i in look_up: look_up[i] += 1
		else: look_up[i] = 1

	for i in str2:
		if i not in look_up: return False
		if look_up[i] == 0: return False
		else: look_up[i] -= 1

	return True
